### PR TITLE
Fix: #37, sync integration

### PR DESCRIPTION
**Pull request by iccub**
Most of the code is a direct 1.6 port, with few small improvements.

Notes:
- I moved `brave/sync` dependency from carthage to npm. It doesn't make sense to place it in Carthage, and it also makes setup script a bit cleaner. See `bootstrap.sh` and `package.json` for changes. Also `brave/crypto` dep is removed, because `brave/sync` relies on it already, no need to duplicate it
- I created a separate folder for soft-linked sync dependencies. Also added a small readme for it.
- Sync callout in Bookmark panel was not added. I will add it in another commit if we want to keep it.
- We continue to utilize `NO_SYNC` flag. It was not added to the project so sync is enabled by default.

### CODE REVIEW

**Comment by iccub** [comment]
Changing to `[WIP]` because tests needs to be updated, but all the main code is ready for review.

**Comment by iccub** [comment]
Tests are fixed, the PR should be ready to go.

It also fixes issues with sorting bookmarks when there is over 10 of them.

File: Client/Frontend/Browser/HomePanel/favorites/FavoritesDataSource.swift [hunk 1]
```diff
@@ -76,7 +76,8 @@ class FavoritesDataSource: NSObject, UICollectionViewDataSource {
             return UICollectionViewCell()
         }
         
-        let ftd = FavoritesTileDecorator(url: url, cell: cell, indexPath: indexPath, color: fav.color != nil ? UIColor(colorString: fav.color!) : nil)
+        // BRAVE TODO: Update Favorites logic, color argument is outdated here. Setting nil for now.b
```
Comment (jhreis) [inline Client/Frontend/Browser/HomePanel/favorites/FavoritesDataSource.swift]: Is there at ticket to track this?
Also, remove extraneous "b"

File: Client/Frontend/Settings/SettingsViewController.swift [hunk 2]
```diff
@@ -156,6 +165,33 @@ class SettingsViewController: TableViewController {
...
+                        
+                        self.navigationController?.pushViewController(syncSettingsVC, animated: true)
+                    } else {
+                        let view = SyncWelcomeViewController()
+                        view.dismissHandler = {
+                            view.navigationController?.popToRootViewController(animated: true)
+                        }
+                        self.navigationController?.pushViewController(view, animated: true)
+                    }
+                    }, accessory: .disclosureIndicator)
```
Comment (jhreis) [inline Client/Frontend/Settings/SettingsViewController.swift]: Fix indentation.

**Review by jhreis — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/HomePanel/favorites/FavoritesDataSource.swift (hunk 1 above)
Comment (iccub) [reply Client/Frontend/Browser/HomePanel/favorites/FavoritesDataSource.swift]: This was fixed in https://github.com/brave/brave-ios/pull/383, I only need to merge latest development changes to this rbanch

**Comment by iccub** [comment]
Fixed a bug with sending empty bookmarks to sync server and added missing strings for Sync(that were removed in the latest localization cleanup)

File: BraveShared/BraveStrings.swift [hunk 3]
```diff
@@ -297,6 +297,58 @@ public extension Strings {
     public static let ErrorPagesVisitOnceButton = NSLocalizedString("ErrorPagesVisitOnceButton", tableName: "BraveShared", value: "Visit site anyway", comment: "Button label to temporarily continue to the site from the certificate error page")
 }
 
+// MARK: - Sync
+public extension Strings {
```
Comment (jhreis) [inline BraveShared/BraveStrings.swift:301]: Please add `value` to all of these strings. For now, the key can just be the variable name. For example:

public static let QRCode = NSLocalizedString("QRCode, value: "QR Code", comment: "QR Code section title")

File: Data/sync/Sync.swift [hunk 4]
```diff
@@ -64,101 +66,138 @@ enum SyncActions: Int {
...
+    
     // TODO: Move to a better place
     fileprivate let prefNameId = "device-id-js-array"
     fileprivate let prefNameName = "sync-device-name"
     fileprivate let prefNameSeed = "seed-js-array"
     fileprivate let prefFetchTimestamp = "sync-fetch-timestamp"
+    fileprivate let prefBaseOrder = "sync-base-order"
     
-//    #if DEBUG
+    //    #if DEBUG
```
Comment (jhreis) [inline Data/sync/Sync.swift]: Make these variables `lazy var` and return based on the build type.

**Review by jhreis — CHANGES_REQUESTED** [review]
Great work. 

Please also file a ticket for converting our `lazy var` usage to using `then`. Do not do it now, just file away so we can track it.

File: Data/sync/Sync.swift [hunk 5]
```diff
@@ -64,101 +66,137 @@ enum SyncActions: Int {
...
     fileprivate let prefNameId = "device-id-js-array"
     fileprivate let prefNameName = "sync-device-name"
     fileprivate let prefNameSeed = "seed-js-array"
     fileprivate let prefFetchTimestamp = "sync-fetch-timestamp"
+    fileprivate let prefBaseOrder = "sync-base-order"
+    
+    fileprivate lazy var isDebug: Bool = { return AppConstants.BuildChannel == .developer }()
+    
+    fileprivate lazy var serverUrl: String = {
+        return AppConstants.BuildChannel == .developer ?
```
Comment (jhreis) [inline Data/sync/Sync.swift]: Use the above `isDebug`, that way logic is centralized:
`return isDebug ? "" : ""`

File: Data/sync/Sync.swift [hunk 6]
```diff
@@ -64,101 +66,137 @@ enum SyncActions: Int {
...
+    }
+    
     // TODO: Move to a better place
     fileprivate let prefNameId = "device-id-js-array"
     fileprivate let prefNameName = "sync-device-name"
     fileprivate let prefNameSeed = "seed-js-array"
     fileprivate let prefFetchTimestamp = "sync-fetch-timestamp"
+    fileprivate let prefBaseOrder = "sync-base-order"
+    
+    fileprivate lazy var isDebug: Bool = { return AppConstants.BuildChannel == .developer }()
```
Comment (jhreis) [inline Data/sync/Sync.swift]: Switch to `isRelease`, so betas also are treated as `debug=true`

**Review by jhreis — APPROVED** [review]

File: Data/sync/Sync.swift [hunk 7]
```diff
@@ -64,101 +66,136 @@ enum SyncActions: Int {
...
+    }
+    
     // TODO: Move to a better place
     fileprivate let prefNameId = "device-id-js-array"
     fileprivate let prefNameName = "sync-device-name"
     fileprivate let prefNameSeed = "seed-js-array"
     fileprivate let prefFetchTimestamp = "sync-fetch-timestamp"
+    fileprivate let prefBaseOrder = "sync-base-order"
+    
+    fileprivate lazy var isDebug: Bool = { return AppConstants.BuildChannel != .release }()
```
Comment (jhreis) [inline Data/sync/Sync.swift]:
```suggestion
    fileprivate lazy var isDebug: Bool = { return !AppConstants.BuildChannel.isRelease }()
```

**Review by jhreis — APPROVED** [review]

**Pull request merged**